core/src/clients/github_poller.rs

```diff
@@ -0,0 +1,73 @@
+use crate::clients::git::{Git, ShouldPrune};
+use crate::clients::github::GraphQLClient;
+use std::path::PathBuf;
+use std::sync::mpsc::Sender as STDSender;
+use std::time::Duration;
+use tracing::{info, span};
+
+pub struct GithubPoller {
+    client: GraphQLClient,
+    owner: String,
+    repo: String,
+    branch: String,
+
+    git: Git,
+    cached_commit: String,
+}
+
+impl GithubPoller {
+    pub async fn new(
+        token: String,
+        owner: String,
+        repo: String,
+        path: String,
+        branch: String,
+        tx: STDSender<String>,
+    ) -> anyhow::Result<Self> {
+        let client = GraphQLClient::new(token).await?;
+
+        let commit = client
+            .get_latest_commit(&owner, &repo, &branch.clone())
+            .await?;
+        info!("Initializing GithubPoller with commit: {}", commit);
+
+        let git = Git::new(PathBuf::from(path.clone()), tx);
+
+        // run an initial fetch
+        git.fetch(ShouldPrune::Yes).await?;
+
+        Ok(GithubPoller {
+            client,
+            owner,
+            repo,
+            branch,
+
+            git,
+            cached_commit: commit,
+        })
+    }
+
+    // poll and take an async fn callback
+    pub async fn poll(&mut self, interval: Duration) -> anyhow::Result<()> {
+        loop {
+            let commit = self
+                .client
+                .get_latest_commit(&self.owner, &self.repo, &self.branch)
+                .await?;

```

Comment:
I feel like I can't tell if this is actually any better than just running `git fetch`. Maybe it's better simply by nature of not messing with the repo all the time.

Reply:
Like why wouldn't we run `git fetch` every 5 seconds?

Reply:
If `git fetch` creates the `index.lock` file, then this change is good since it's really annoying as a developer to run index-locking git operations in the repo directly only to have it error out that something else is locking the repo.

Reply:
`fetch` doesn't touch index.lock - chris and i have been discussing and we think there's a better way than this

Reply:
I'm gunna get rid of the github API stuff and instead set up background tasks for fetching and repo maintenance. i think we can get what we want that way.